### PR TITLE
Add launcher options allowing to override the default Scala version

### PR DESCRIPTION
**Pull request by Gedochao**
Relevant to #2838 

This PR adds 2 new launcher options: `--cli-default-scala-version` & `--cli-predefined-repository`

`--cli-default-scala-version` allows to override the default Scala version (which can then be overridden with `-S`):
```bash
scala-cli --cli-default-scala-version 3.4.2-RC1 version --offline
# Scala CLI version: 1.2.3-SNAPSHOT
# Scala version (default): 3.4.2-RC1
scala-cli --cli-default-scala-version 3.4.2-RC1 run -e 'println(dotty.tools.dotc.config.Properties.simpleVersionString)' --with-compiler
# Compiling project (Scala 3.4.2-RC1, JVM (17))
# Compiled project (Scala 3.4.2-RC1, JVM (17))
# 3.4.2-RC1
scala-cli --cli-default-scala-version 3.4.2-RC1 run -e 'println(dotty.tools.dotc.config.Properties.simpleVersionString)' --with-compiler -S lts
# Compiling project (Scala 3.3.3, JVM (17))
# Compiled project (Scala 3.3.3, JVM (17))
# 3.3.3
```

`--cli-predefined-repository` is an equivalent to the existing `--repo` option for building sub-commands, but can be passed as a launcher option. This i.e. allows to pass a local repository containing compiler artifacts.

```bash
cs fetch --cache local-repo org.scala-lang:scala3-compiler_3:3.4.2-RC1
# local-repo/https/repo1.maven.org/maven2/org/scala-lang/scala3-compiler_3/3.4.1-RC1/scala3-compiler_3-3.4.1-RC1.jar
# local-repo/https/repo1.maven.org/maven2/org/scala-lang/scala3-interfaces/3.4.1-RC1/scala3-interfaces-3.4.1-RC1.jar
# local-repo/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.4.1-RC1/scala3-library_3-3.4.1-RC1.jar
# local-repo/https/repo1.maven.org/maven2/org/scala-lang/tasty-core_3/3.4.1-RC1/tasty-core_3-3.4.1-RC1.jar
# local-repo/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-asm/9.6.0-scala-1/scala-asm-9.6.0-scala-1.jar
# local-repo/https/repo1.maven.org/maven2/org/scala-sbt/compiler-interface/1.9.6/compiler-interface-1.9.6.jar
# local-repo/https/repo1.maven.org/maven2/org/jline/jline-reader/3.19.0/jline-reader-3.19.0.jar
# local-repo/https/repo1.maven.org/maven2/org/jline/jline-terminal/3.19.0/jline-terminal-3.19.0.jar
# local-repo/https/repo1.maven.org/maven2/org/jline/jline-terminal-jna/3.19.0/jline-terminal-jna-3.19.0.jar
# local-repo/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.12/scala-library-2.13.12.jar
# local-repo/https/repo1.maven.org/maven2/org/scala-sbt/util-interface/1.9.8/util-interface-1.9.8.jar
# local-repo/https/repo1.maven.org/maven2/net/java/dev/jna/jna/5.3.1/jna-5.3.1.jar
scala-cli --cli-default-scala-version 3.4.2-RC1 --cli-predefined-repository file:///Users/pchabelski/IdeaProjects/scala-cli-tests/demo/local-repo/https/repo1.maven.org/maven2 run simple.sc --with-compiler --offline --power
# 3.4.2-RC1
```

### CODE REVIEW

**Comment by Gedochao** [comment]
cc @bishabosha

**Comment by bishabosha** [comment]
nice, Is it necessary to add another test to assert it fails if --cli-default-scala-version is provided twice?

File: modules/options/src/main/scala/scala/build/options/BuildOptions.scala [hunk 1]
```diff
@@ -247,7 +247,12 @@ final case class BuildOptions(
 
   // used when downloading fails
   private def defaultStableScalaVersions =
-    Seq(defaultScala212Version, defaultScala213Version, defaultScalaVersion)
+    Seq(
+      defaultScala212Version,
+      defaultScala213Version,
+      defaultScalaVersion,
+      scalaOptions.defaultScalaVersion.getOrElse(defaultScalaVersion)
+    ).distinct
```
Comment (MaciejG604) [inline modules/options/src/main/scala/scala/build/options/BuildOptions.scala]: This whole variable seems to be unused, I'd delete it

**Review by MaciejG604 — APPROVED** [review]

File: modules/options/src/main/scala/scala/build/options/BuildOptions.scala (hunk 1 above)
Comment (Gedochao) [reply modules/options/src/main/scala/scala/build/options/BuildOptions.scala]: Oh, you seem to be correct.
Thanks, deleted

**Comment by Gedochao** [comment]
> Is it necessary to add another test to assert it fails if --cli-default-scala-version is provided twice?

@bishabosha that's guaranteed by `case-app`, but I added a test as a sanity check

**Review by kasiaMarek — COMMENTED** [review]
Generally looks good, but there are two places, where this doesn't seem to be propagated and I don't know if that is on purpose:
1. scala version used for `export`,
2. when scala version is downgraded for ammonite it's check against `defaultScalaVersion` not the override default

**Comment by Gedochao** [comment]
> Generally looks good, but there are two places, where this doesn't seem to be propagated and I don't know if that is on purpose:
> 
> 1. scala version used for `export`,
> 2. when scala version is downgraded for ammonite it's check against `defaultScalaVersion` not the override default

@kasiaMarek nice catch, let me fix those

**Comment by Gedochao** [comment]
Changes:
- export `--sbt`, `--mill` and `--json` should now respect the launcher override for the default Scala version
- ammonite repl should now respect the launcher override when downgrading the default Scala version

**Review by kasiaMarek — APPROVED** [review]

**Comment by bishabosha** [comment]
nice

**Pull request merged**